// react-spatial/components/Search/engines/StopFinder.js
import React from 'react';

import { StopsAPI } from 'mobility-toolbox-js/api';
import Engine from './Engine';

var StopFinder = /*@__PURE__*/(function (Engine) {
  function StopFinder(endpoint, options) {
    if ( endpoint === void 0 ) endpoint = 'https://api.geops.io/stops/v1/';
    if ( options === void 0 ) options = {};

    Engine.call(this);
    this.options = options;
    // apiKey is passed as key in StopsAPI, delete to avoid duplicates.
    delete this.options.apiKey;
    this.api = new StopsAPI({
      url: endpoint,
      apiKey: options.apiKey,
    });
  }

  if ( Engine ) StopFinder.__proto__ = Engine;
  StopFinder.prototype = Object.create( Engine && Engine.prototype );
  StopFinder.prototype.constructor = StopFinder;

  StopFinder.prototype.search = function search (value) {
    return this.api.search(Object.assign({}, {q: encodeURIComponent(value)},
      this.options));
  };

  StopFinder.prototype.setApiKey = function setApiKey (apiKey) {
    this.api.apiKey = apiKey;
  };

  StopFinder.prototype.render = function render (item) {
    return React.createElement( 'div', null, item.properties.name );
  };

  StopFinder.value = function value (item) {
    return item.properties.name;
  };

  return StopFinder;
}(Engine));

export default StopFinder;
