// @geops/geops-ui/themes/geops-ui.js
import { createMuiTheme } from '@material-ui/core/styles';

export var colors = {
  buttonDisabledGray: '#707070',
  outlinedButtonGreen: '#DEF1CA',
  primaryBlack: '#353535',
  primaryBlue: '#6987a1',
  primaryDarkBlue: '#4B6881',
  primaryDarkGreen: '#549413',
  primaryGreen: '#76B833',
  primaryLightBlue: '#7FA8CC',
  primaryLightGreen: '#A1D66A',
  secondaryDarkGray: '#BEBEBE',
  secondaryDarkOrange: '#D59734',
  secondaryDarkRed: '#98271D',
  secondaryGray: '#EFEFEF',
  secondaryLightGray: '#FCFCFC',
  secondaryLightOrange: '#EBBF7B',
  secondaryLightRed: '#C8716A',
  secondaryOrange: '#E5AC52',
  secondaryRed: '#A94941',
  white: '#FFFFFF',
  footer: '#333333',
  msDarkBlue: '#1789C9',
  msLightBlue: '#63B3F4',
  msSecondaryGray: '#F9F9F9',
};

export var muiTheme = createMuiTheme({
  colors: colors,
  props: {
    MuiButtonBase: {
      disableRipple: true,
    },
  },
  typography: {
    fontFamily: ['Lato', 'Arial'],
    fontWeight: 400, // Regular, see https://fonts.google.com/specimen/Lato
    h1: {
      fontSize: 50,
      lineHeight: '60px',
      fontWeight: 900, // Black
      marginTop: 5,
      marginBottom: 5,
    },
    h2: {
      fontSize: 36,
      lineHeight: '44px',
      fontWeight: 900, // Black
      marginTop: 10,
      marginBottom: 10,
    },
    h3: {
      fontSize: 22,
      lineHeight: '27px',
      fontWeight: 900, // Black
      marginTop: 10,
      marginBottom: 10,
    },
    body1: {
      fontSize: 18,
      lineHeight: '24px',
    },
    button: {
      fontSize: 18,
      fontWeight: 700, // Bold
      lineHeight: '24px',
      textTransform: 'none',
    },
  },
  palette: {
    primary: {
      main: colors.primaryBlue,
    },
    secondary: {
      main: colors.primaryGreen,
    },
    text: {
      primary: colors.primaryBlack,
    },
  },
  overrides: {
    MuiTypography: {
      root: {
        '&.headline': {
          paddingBottom: 5,
          position: 'relative',
          '&::before': {
            content: '""',
            position: 'absolute',
            width: 95,
            bottom: 0,
            left: 0,
            borderBottomWidth: 4,
            borderBottomStyle: 'solid',
            borderBottomColor: colors.primaryGreen,
            borderRadius: 5,
            marginLeft: 0,
            marginBottom: -8,
          },
        },
      },
      h1: {
        '&.headline': {
          marginBottom: 40,
        },
      },
      h2: {
        '&.headline': {
          marginBottom: 30,
        },
      },
      h3: {
        '&.headline': {
          marginBottom: 30,
        },
      },
    },
    MuiAppBar: {
      root: {
        display: 'flex',
        justifyContent: 'space-between',
        flexDirection: 'row',
        minHeight: 100,
        borderBottomWidth: 4,
        borderBottomStyle: 'solid',
        borderBottomColor: colors.primaryBlue,
      },
    },
    MuiCard: {
      root: {
        height: '100%',
      },
    },
    MuiCardContent: {
      root: {
        padding: '16px 32px',
      },
    },
    MuiCardActionArea: {
      root: {
        '&:hover .MuiCardActionArea-focusHighlight': {
          opacity: 0,
        },
      },
    },
    MuiCardActions: {
      root: {
        padding: 0,
      },
    },
    MuiList: {
      root: {
        color: colors.primaryBlack,
        fontFamily: 'Lato, Arial, sans-serif',
      },
    },
    MuiTabs: {
      root: {
        height: '100%',
        color: colors.primaryBlack,
      },
      flexContainer: {
        height: '100%',
      },
      indicator: {
        backgroundColor: colors.primaryGreen,
      },
    },
    MuiTab: {
      root: {
        textTransform: 'unset',
        '@media (min-width: 600px)': {
          minWidth: 'unset',
        },
        width: 100,
        fontFamily: 'Lato, Arial, sans-serif',
        color: colors.primaryBlue,
        '& .MuiButtonBase-root': {
          color: colors.primaryBlue,
        },
      },
      textColorInherit: {
        opacity: 1,
      },
    },
    MuiTableCell: {
      head: {
        fontSize: 22,
        fontWeight: 900, // Black
      },
    },
    MuiIconButton: {
      root: {
        color: colors.primaryBlue,
      },
    },
    MuiButton: {
      root: {
        backgroundColor: colors.primaryGreen,
        color: colors.white,
        borderRadius: 5,
        '&:hover': {
          backgroundColor: colors.primaryGreen,
          boxShadow: '0px 5px 16px #35353540',
        },
        '&:disabled': {
          backgroundColor: colors.secondaryDarkGray,
          color: colors.buttonDisabledGray,
        },
        '&.v2': {
          borderWidth: 2,
          borderStyle: 'solid',
          borderColor: colors.primaryGreen,
          color: colors.primaryGreen,
          backgroundColor: colors.outlinedButtonGreen,
          '&:hover': {
            backgroundColor: colors.primaryGreen,
            boxShadow: '0px 5px 16px #35353540',
            color: colors.white,
          },
          '&:disabled': {
            color: colors.buttonDisabledGray,
            backgroundColor: colors.secondaryGray,
            borderColor: colors.secondaryDarkGray,
          },
        },
        '&.large': {
          minWidth: 256,
          height: 49,
        },
      },
      text: {
        padding: '10px 25px',
      },
    },
  },
});

export default muiTheme;
