// @geops/geops-ui/components/Footer.js
import React from 'react';
import PropTypes from 'prop-types';
import { v1 as uniqueKey } from 'uuid';
import { IconButton, Link } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import FooterLogo from '../images/icons/footer_logo.svg';
import TwitterIcon from '../images/icons/twitter.svg';
import FacebookIcon from '../images/icons/facebook.svg';
import LinkedInIcon from '../images/icons/linkedin.svg';
import GitHubIcon from '../images/icons/github.svg';
import XingIcon from '../images/icons/xing.svg';

var propTypes = {
  /**
   * List tabs configuration.
   */
  links: PropTypes.arrayOf(
    PropTypes.shape({
      /**
       * The link label.
       */
      label: PropTypes.string.isRequired,

      /**
       * href
       */
      href: PropTypes.string,
    })
  ),
};

var defaultProps = {
  links: [
    {
      label: 'About geOps',
      href: 'https://geops.ch/about?language=en',
    },
    {
      label: 'Privacy Policy',
      href: 'https://geops.ch/datenschutz',
    },
    {
      label: 'Imprint',
      href: 'https://geops.ch/impressum',
    } ],
};

var useStyles = makeStyles(function (theme) {
  var obj, obj$1;

  return ({
  footer: ( obj = {
    bottom: 0,
    left: 0,
    right: 0,
    display: 'flex',
    flexWrap: 'wrap',
    alignItems: 'center',
    justifyContent: 'space-between',
    backgroundColor: theme.colors.footer,
    padding: '0 30px',
    minHeight: 70,
    zIndex: 1100
  }, obj[theme.breakpoints.down('sm')] = {
      minHeight: 100,
      flexDirection: 'column',
      alignItems: 'initial',
    }, obj ),
  footerElement: {
    display: 'flex',
    alignItems: 'center',
    margin: '15px 0',
  },
  brand: {
    minWidth: 200,
  },
  logo: {
    padding: 10,
    width: 80,
  },
  icon: {
    padding: 10,
  },
  svg: {
    width: 20,
    height: 20,
  },
  links: ( obj$1 = {}, obj$1[theme.breakpoints.down('sm')] = {
      display: 'flex',
      flexDirection: 'column',
    }, obj$1 ),
  link: {
    marginLeft: 10,
    marginRight: 10,
    whiteSpace: 'nowrap',
    color: 'white',
    fontSize: 16,
  },
});
});

var Footer = function (ref) {
  var links = ref.links;

  var classes = useStyles();
  return (
    React.createElement( 'footer', { className: classes.footer },
      React.createElement( 'div', { className: ((classes.footerElement) + " " + (classes.brand)) },
        React.createElement( 'img', { className: classes.logo, src: FooterLogo, alt: "Logo" })
      ),
      React.createElement( 'div', { className: classes.links },
        Array.isArray(links) &&
          links.map(function (link) { return (
            React.createElement( Link, {
              key: uniqueKey(), href: link.href, className: classes.link, variant: "body1" },
              link.label
            )
          ); })
      ),
      React.createElement( 'div', { className: ("" + (classes.footerElement)) },
        React.createElement( IconButton, {
          className: classes.icon, href: "https://twitter.com/geops?lang=en", target: "_blank", rel: "noopener" },
          React.createElement( 'img', { className: classes.svg, src: TwitterIcon, alt: "twitter" })
        ),
        React.createElement( IconButton, {
          className: classes.icon, href: "https://www.facebook.com/geOpsSpatialWeb", target: "_blank", rel: "noopener" },
          React.createElement( 'img', { className: classes.svg, src: FacebookIcon, alt: "facebook" })
        ),
        React.createElement( IconButton, {
          className: classes.icon, href: "https://github.com/geops/", target: "_blank", rel: "noopener" },
          React.createElement( 'img', { className: classes.svg, src: GitHubIcon, alt: "github" })
        ),
        React.createElement( IconButton, {
          className: classes.icon, href: "https://de.linkedin.com/company/geops", target: "_blank", rel: "noopener" },
          React.createElement( 'img', { className: classes.svg, src: LinkedInIcon, alt: "linkedin" })
        ),
        React.createElement( IconButton, {
          className: classes.icon, href: "https://www.xing.com/companies/geops", target: "_blank", rel: "noopener" },
          React.createElement( 'img', { className: classes.svg, src: XingIcon, alt: "xing" })
        )
      )
    )
  );
};

Footer.propTypes = propTypes;
Footer.defaultProps = defaultProps;

export default Footer;
